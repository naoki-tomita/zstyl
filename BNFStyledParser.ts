type Types =
    "StyleSheet"
  | "LocalStyle"
  | "Style"
  | "Identifier"
  | "Value"
  | "Values"
  | "NestedStyle"
  | "Selector"
  | "Block"
  | "MediaCondition"
  | "Media"
  | "Keyframe"
  ;

interface Ast {
  type: Types;
};

type RootStyle = LocalStyle | NestedStyle | MediaStyle | KeyframeStyle;

export interface StyleSheetAst extends Ast {
  type: "StyleSheet";
  children: Array<LocalStyle | NestedStyle | MediaStyle | KeyframeStyle>;
}

export const StyleSheetParser: Parser<StyleSheetAst> = {
  parse(style: string): ParsedResult<StyleSheetAst> {
    const children: StyleSheetAst["children"] = [];
    let result: ParsedResult<RootStyle> = { remaining: style };
    let successToParse;
    do {
      successToParse = false;
      for (const parser of [LocalStyleParser, NestedStyleParser, MediaStyleParser, KeyframeStyleParser]) {
        result = parser.parse(result.remaining);
        if (result.ast) {
          children.push(result.ast);
          successToParse = true;
        }
      }
    } while (successToParse)

    return {
      ast: {
        type: "StyleSheet",
        children,
      },
      remaining: result.remaining
    };
  }
}

export interface Identifier extends Ast {
  type: "Identifier";
  name: string;
}

export const IdentifierParser: Parser<Identifier> = {
  parse(style: string): ParsedResult<Identifier> {
    style = style.trimStart();
    const result = style.match(/^([a-zA-Z0-9\-_.]+)/);
    if (result == null || result[0].length === 0) {
      return { remaining: style };
    }
    const name = result[0];
    return {
      ast: {
        type: "Identifier",
        name,
      },
      remaining: style.slice(name.length),
    }
  }
}

export interface Value extends Ast {
  type: "Value";
  value: string;
}

export const ValueParser: Parser<Value> = {
  parse(style: string): ParsedResult<Value> {
    const result = style.trimStart().match(/^([a-zA-Z0-9\-%#_.\()]+)/);
    if (result == null || result[0].length === 0) {
      return { remaining: style };
    }
    const value = result[0];
    return {
      ast: {
        type: "Value",
        value,
      },
      remaining: style.trimStart().slice(value.length),
    }
  }
}

export interface Values extends Ast {
  type: "Values";
  values: Value[];
}

export const ValuesParser: Parser<Values> = {
  parse(style: string): ParsedResult<Values> {
    let result = ValueParser.parse(style);
    if (!result.ast) {
      return { remaining: style };
    }
    const results: Value[] = [result.ast];
    while (true) {
      const remaining = result.remaining;
      result = ValueParser.parse(remaining);
      if (!result.ast) {
        return {
          ast: {
            type: "Values",
            values: results,
          },
          remaining,
        }
      }
      results.push(result.ast);
    }
  }
}

export interface Style extends Ast {
  type: "Style"
  prop: Identifier;
  values: Values;
}
export const StyleParser: Parser<Style> = {
  parse(style: string): ParsedResult<Style> {
    const identifier = IdentifierParser.parse(style);
    if (!identifier.ast) {
      return { remaining: style };
    }
    const colonWithIdentifiers = identifier.remaining.trimStart();
    if (!colonWithIdentifiers.startsWith(":")) {
      return { remaining: style };
    }
    const identifiers = ValuesParser.parse(colonWithIdentifiers.slice(1));
    if (!identifiers.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Style",
        prop: identifier.ast,
        values: identifiers.ast,
      },
      remaining: identifiers.remaining
    }
  }
}

export interface LocalStyle extends Ast {
  type: "LocalStyle";
  prop: Identifier;
  values: Values;
}
export const LocalStyleParser: Parser<LocalStyle> = {
  parse(style: string): ParsedResult<LocalStyle> {
    const result = StyleParser.parse(style);
    if (
      !result.ast ||
      !result.remaining.trimStart().startsWith(";")
    ) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "LocalStyle",
        prop: result.ast.prop,
        values: result.ast.values,
      },
      remaining: result.remaining.trimStart().slice(1)
    }
  }
}

type AnyStyle = LocalStyle | NestedStyle;

export interface Selector extends Ast {
  type: "Selector";
  value: string; // ここを BNF で表現するのはあきらめました
}

export const SelectorParser: Parser<Selector> = {
  parse(style: string): ParsedResult<Selector> {
    const result = style.trimStart().match(/^([a-zA-Z0-9_()+>| ,.#~=%^$&\[\]"'*:/\-]+)/)
    if (result == null) {
      return { remaining: style };
    }
    const value = result[0].trim()
    return {
      ast: {
        type: "Selector",
        value,
      },
      remaining: style.trimStart().slice(value.length)
    }
  }
}

export interface Block extends Ast {
  type: "Block";
  body: AnyStyle[];
}

export const BlockParser: Parser<Block> = {
  parse(style: string): ParsedResult<Block> {
    if (!style.trimStart().startsWith("{")) {
      return { remaining: style };
    }
    let result: ParsedResult<AnyStyle> = { remaining: style.trimStart().slice(1) };
    const results: AnyStyle[] = [];
    let successToParse: boolean;
    do {
      successToParse = false;
      for (const parser of [LocalStyleParser, NestedStyleParser]) {
        result = parser.parse(result.remaining);
        if (result.ast) {
          successToParse = true;
          results.push(result.ast);
        }
      }
    } while (successToParse)
    if (!result.remaining.trimStart().startsWith("}")) {
      return { remaining: style }
    }
    return {
      ast: {
        type: "Block",
        body: results,
      },
      remaining: result.remaining.trimStart().slice(1)
    }
  }
}

export interface NestedStyle extends Ast {
  type: "NestedStyle";
  selector: Selector;
  block: Block;
}

export const NestedStyleParser: Parser<NestedStyle> = {
  parse(style: string): ParsedResult<NestedStyle> {
    const selector = SelectorParser.parse(style);
    if (!selector.ast) {
      return { remaining: style };
    }
    const block = BlockParser.parse(selector.remaining);
    if (!block.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "NestedStyle",
        selector: selector.ast,
        block: block.ast,
      },
      remaining: block.remaining
    }
  }
}

interface MediaCondition extends Ast {
  type: "MediaCondition";
  name: Identifier;
  value: Identifier;
}

export interface MediaStyle extends Ast {
  type: "Media"
  condition: MediaCondition;
  block: Block;
}

export const MediaStyleParser: Parser<MediaStyle> = {
  parse(style: string): ParsedResult<MediaStyle> {
    if (!style.trimStart().startsWith("@media")) {
      return { remaining: style };
    }
    const slicedAtMediaStyle = style.trimStart().slice("@media".length);
    if (!slicedAtMediaStyle.trimStart().startsWith("(")) {
      return { remaining: style };
    }
    const leftParenTrimmedStyle = slicedAtMediaStyle.trimStart().slice(1);
    const nameResult = IdentifierParser.parse(leftParenTrimmedStyle);
    if (!nameResult.ast) {
      return { remaining: style };
    }
    if (!nameResult.remaining.trimStart().startsWith(":")) {
      return { remaining: style };
    }
    const valueResult = IdentifierParser.parse(nameResult.remaining.trimStart().slice(1));
    if (!valueResult.ast) {
      return { remaining: style };
    }
    if (!valueResult.remaining.trimStart().startsWith(")")) {
      return { remaining: style };
    }
    const rightParenTrimmedStyle = valueResult.remaining.trimStart().slice(1);
    const blockResult = BlockParser.parse(rightParenTrimmedStyle);
    if (!blockResult.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Media",
        condition: {
          type: "MediaCondition",
          name: nameResult.ast,
          value: valueResult.ast,
        },
        block: blockResult.ast
      },
      remaining: blockResult.remaining
    }
  }
}

export interface KeyframeStyle extends Ast {
  type: "Keyframe"
  name: Identifier;
  block: Block;
}

export const KeyframeStyleParser: Parser<KeyframeStyle> = {
  parse(style: string): ParsedResult<KeyframeStyle> {
    if (!style.trimStart().startsWith("@keyframe")) {
      return { remaining: style };
    }
    const keyframeExcludedStyle = style.trimStart().slice("@keyframe".length);
    const nameResult = IdentifierParser.parse(keyframeExcludedStyle);
    if (!nameResult.ast) {
      return { remaining: style };
    }
    const blockStyle = nameResult.remaining;
    const blockResult = BlockParser.parse(blockStyle);
    if (!blockResult.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Keyframe",
        name: nameResult.ast,
        block: blockResult.ast
      },
      remaining: blockResult.remaining,
    }
  }
}

type ParsedResult<T extends Ast> = {
  ast?: T, remaining: string
};

interface Parser<T extends Ast> {
  parse(style: string): ParsedResult<T>;
}
