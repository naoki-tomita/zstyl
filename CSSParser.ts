
type MediaQueryParseResult = {
  startIndex: number;
  lastIndex: number;
  body: string;
  rule: string;
  fullStyle: string;
};

type NestedStyleParseResult = {
  startIndex: number;
  lastIndex: number;
  body: string;
  querySelector: string;
  fullStyle: string;
}

function space(width: number) {
  return Array(width).fill(null).map(() => " ").join("");
}
function indent(text: string, width: number) {
  return text.split("\n").map(t => `${space(width)}${t}`).join(`\n`);
}
export class CSSParser {
  readonly style: string;
  minify(style: string): string {
    return style.replace(/\n/g, "").replace(/ {2,}/g, " ").trim();
  }

  constructor(style: string) {
    this.style = this.minify(style);
  }

  fillWithId(id: string) {
    let style = this.style;
    const results: string[] = [];
    for (const mediaQuery of this.getMediaQueries(style)) {
      const parser = new CSSParser(mediaQuery.body);
      const filled = parser.fillWithId(id);
      style = style.replace(mediaQuery.fullStyle, "");
      results.push(`@media (${mediaQuery.rule}) { \n${indent(filled, 2)}\n}`);
    }
    for (const nestedStyle of this.getNestedStyles(style)) {
      const parser = new CSSParser(nestedStyle.body);
      const filled = parser.fillWithId(nestedStyle.querySelector.startsWith("&") ? nestedStyle.querySelector.replace("&", id) : `${id} ${nestedStyle.querySelector}`);
      style = style.replace(nestedStyle.fullStyle, "");
      results.push(filled);
    }
    results.unshift(`${id} { ${style.trim()} }`);
    return results.join("\n");
  }

  getMediaQueries(style: string = this.style): MediaQueryParseResult[] {
    const results: MediaQueryParseResult[] = [];
    const parser = new Parser(style);
    let startIndex = parser.findIndexAfter("@media");
    while (startIndex !== -1) {
      let lastIndex = startIndex;
      const firstBraceIndexOf = parser.peek(startIndex).findIndexAfter("{");
      parser.peek(firstBraceIndexOf);
      let leftBrace = 0;
      let rightBrace = 0;
      while (true) {
        const char = parser.getOne();
        if (char === "{") leftBrace++;
        if (char === "}") rightBrace++;
        if (leftBrace === rightBrace) {
          lastIndex = parser.currentIndex();
          break;
        }
      }
      results.push({
        startIndex,
        lastIndex,
        body: parser.get(firstBraceIndexOf + 1, lastIndex - 1).trim(),
        rule: parser.get(startIndex, firstBraceIndexOf).match(/@media\s*\((.+)\)/)?.[1] ?? "",
        fullStyle: style.slice(startIndex, lastIndex),
      })
      startIndex = parser.findIndexAfter("@media");
    }
    return results;
  }

  getNestedStyles(style: string = this.style): NestedStyleParseResult[] {
    const results: NestedStyleParseResult[] = []
    const parser = new Parser(style);
    let startIndex = parser.findIndexAfter("{");
    while (startIndex !== -1) {
      const startIndexOfQuerySelector = parser.peek(startIndex).findLastIndexBefore("}", ";") + 1;
      const querySelector = parser.get(startIndexOfQuerySelector + 1, startIndex - 1).trim();
      let leftBrace = 0;
      let rightBrace = 0;
      let lastIndex = startIndexOfQuerySelector;
      while (true) {
        const currnet = parser.getOne();
        if (currnet === "{") leftBrace++;
        if (currnet === "}") rightBrace++;
        if (leftBrace === rightBrace) {
          lastIndex = parser.currentIndex();
          break;
        }
      }
      results.push({
        startIndex: startIndexOfQuerySelector,
        lastIndex,
        querySelector,
        body: parser.get(startIndex + 1, lastIndex - 1).trim(),
        fullStyle: parser.get(startIndexOfQuerySelector, lastIndex).trim(),
      });
      startIndex = parser.peek(lastIndex).findIndexAfter("{");
    }
    return results;
  }
}

class Parser {
  index: number = 0;
  constructor(readonly text: string) {}

  next(): string {
    return this.text[++this.index];
  }
  currentIndex(): number {
    return this.index;
  }
  current() {
    return this.text[this.index];
  }
  getOne() {
    return this.text[this.index++];
  }
  peek(index: number) {
    this.index = index;
    return this;
  }
  findLastIndexBefore(...texts: string[]) {
    return Math.max(...texts.map(text => this.text.slice(0, this.index).lastIndexOf(text)), 0);
  }
  findIndexAfter(...texts: string[]) {
    const indexOf = Math.min(...texts.map(text => this.text.slice(this.index).indexOf(text)), Infinity);
    return indexOf === -1 ? -1 : this.index + indexOf;
  }
  get(from: number, to: number) {
    return this.text.slice(from, to);
  }
}
