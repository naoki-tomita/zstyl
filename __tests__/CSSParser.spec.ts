import { StyleSheetParser } from "../BNFStyledParser";
import { CSSParser } from "../CSSParser";

describe("CSSParser", () => {
  describe('fillWithId', () => {
    it("should parse style and fill id.", () => {
      const style = `
        display: flex;

        &:hover {
          color: red;
        }

        .wrapper {
          color: blue;
        }

        @media (max-width: 200px) {
          display: inline-flex;
        }
      `;
      const parser = new CSSParser(style);
      expect(parser.fillWithId("#element")).toMatchSnapshot();
    });

    it("should parse and fill id for multiple nested style.", () => {
      const style = `
        display: flex;

        &:hover {
          color: red;

          .inner {
            color: black;
          }
        }

        .wrapper {
          color: blue;

          .inner {
            color: green;
          }
        }

        @media (max-width: 200px) {
          display: inline-flex;

          .inner {
            color: yellow;
          }
        }
      `;
      const parser = new CSSParser(style);
      expect(parser.fillWithId("#element")).toMatchSnapshot();
    });

    it("should real pattern", () => {
      const style = `
        display: flex;
        gap: 16px;
        justify-content: center;

        @media (max-width: 940px) {
          flex-direction: column;
        }
      `;
      const parser = new CSSParser(style);
      expect(parser.fillWithId("*[data-zstyl='aaaaa']")).toMatchSnapshot();
    });
  });

  describe("#renderLocalStylesWithId", () => {
    const tests = [
      // `
      //   display: flex;
      //   background: black center;
      // `,
    ]

    const parser = new CSSParser("");
    tests.forEach((style) => {
      it("should render successfully", () => {
        const result = StyleSheetParser.parse(style);
        expect(parser.renderLocalStylesWithId("id", result.ast?.children as any)).toMatchSnapshot();
      });
    })
  });

});
