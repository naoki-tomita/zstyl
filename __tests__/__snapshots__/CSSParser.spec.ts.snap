// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSSParser fillWithId should parse and fill id for multiple nested style. 1`] = `
"#element { display: flex; }
@media (max-width: 200px) {
  #element { display: inline-flex; }
  #element .inner { color: yellow; }
}
#element:hover { color: red; }
#element:hover .inner { color: black; }
#element .wrapper { color: blue; }
#element .wrapper .inner { color: green; }"
`;

exports[`CSSParser fillWithId should parse style and fill id. 1`] = `
"#element { display: flex; }
@media (max-width: 200px) {
  #element { display: inline-flex; }
}
#element:hover { color: red; }
#element .wrapper { color: blue; }"
`;

exports[`CSSParser fillWithId should real pattern 1`] = `
"*[data-zstyl='aaaaa'] { display: flex; gap: 16px; justify-content: center; }
@media (max-width: 940px) {
  *[data-zstyl='aaaaa'] { flex-direction: column; }
}"
`;

exports[`CSSParser getMediaQueries should parse media query 1`] = `
[
  {
    "body": "display: inline-flex;",
    "fullStyle": "@media (max-width: 720px) { display: inline-flex; }",
    "lastIndex": 66,
    "rule": "max-width: 720px",
    "startIndex": 15,
  },
]
`;

exports[`CSSParser getNestedStyles should parse multiple nested styles 1`] = `
[
  {
    "body": "display: inline-flex; color: black;",
    "fullStyle": ".wrapper1 { display: inline-flex; color: black; }",
    "lastIndex": 64,
    "querySelector": ".wrapper1",
    "startIndex": 14,
  },
  {
    "body": "display: block; color: black;",
    "fullStyle": ".wrapper2 { display: block; color: black; }",
    "lastIndex": 108,
    "querySelector": ".wrapper2",
    "startIndex": 64,
  },
]
`;

exports[`CSSParser getNestedStyles should parse multiple styles 1`] = `
[
  {
    "body": "display: inline-flex; color: black;",
    "fullStyle": ".wrapper { display: inline-flex; color: black; }",
    "lastIndex": 63,
    "querySelector": ".wrapper",
    "startIndex": 14,
  },
]
`;

exports[`CSSParser getNestedStyles should parse nested style 1`] = `
[
  {
    "body": "display: inline-flex;",
    "fullStyle": ".wrapper { display: inline-flex; }",
    "lastIndex": 49,
    "querySelector": ".wrapper",
    "startIndex": 14,
  },
]
`;

exports[`CSSParser getNestedStyles should parse pseudo class 1`] = `
[
  {
    "body": "display: block;",
    "fullStyle": "&:hover { display: block; }",
    "lastIndex": 42,
    "querySelector": "&:hover",
    "startIndex": 14,
  },
]
`;
